zircon-object/src/vm/stream.rs

```diff
@@ -0,0 +1,265 @@
+        if overflow {
+            return Err(ZxError::FILE_BIG);
+        }
+        if target_size > content_size {
+            content_size = self.vmo.set_content_size_and_resize(target_size, offset)?;
+        }
+        if offset >= content_size {
+            return Err(ZxError::NO_SPACE);
+        }
+        let length = count.min(content_size - offset);
+        self.vmo.write(offset, data)?;
+        // unsafe code can be avoided by adding more interface to VmObject and HAL
+        let io_vec = unsafe { slice::from_raw_parts(data.as_ptr(), length) };
+        self.vmo.write(offset, io_vec)?;
+        Ok(length)
+    }
+
+    /// Modify the current seek offset of the stream
+    pub fn seek(&self, seek_origin: SeekOrigin, offset: isize) -> ZxResult<usize> {
+        let mut seek = self.seek.lock();
+        match seek_origin {
+            SeekOrigin::SeekOriginStart => {
+                if offset < 0 {
+                    return Err(ZxError::INVALID_ARGS);
+                }
+                *seek = offset as usize;
+            }
+            SeekOrigin::SeekOriginCurrent => {
+                if offset >= 0 {
+                    let (target, overflow) = seek.overflowing_add(offset as usize);
+                    if overflow {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target;
+                } else {
+                    let target = *seek as isize + offset;
+                    if *seek as isize > 0 && target < 0 {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target as usize;
+                }
+            }
+            SeekOrigin::SeekOriginEnd => {
+                let content_size = self.vmo.content_size();
+                if offset >= 0 {
+                    let (target, overflow) = content_size.overflowing_add(offset as usize);
+                    if overflow {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target;
+                } else {
+                    let target = content_size as isize + offset;
+                    if content_size as isize > 0 && target < 0 {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target as usize;
+                }
+            }
+            _ => unreachable!(),
+        }
+        Ok(*seek)
+    }
+
+    /// Get information of the socket.
+    pub fn get_info(&self) -> StreamInfo {
+        let seek = self.seek.lock();
+        StreamInfo {
+            options: self.options.bits,
+            padding1: 0,
+            seek: *seek as u64,
+            content_size: self.vmo.content_size() as u64,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct StreamInfo {
+    options: u32,
+    padding1: u32,
+    seek: u64,
+    content_size: u64,
+}
```

Comment:
Append docs here from Zircon source code.
(Because I'm confused with the concept of `content_size`...